cborg/src/Codec/CBOR/ByteArray/Sliced.hs

```diff
@@ -0,0 +1,140 @@
+-- Copyright   : (c) Ben Gamari 2017-2018
+-- License     : BSD3-style (see LICENSE.txt)
+--
+-- Maintainer  : duncan@community.haskell.org
+-- Stability   : experimental
+-- Portability : non-portable (GHC extensions)
+--
+-- A ByteArray with more instances than 'Data.Primitive.ByteArray.ByteArray'.
+-- Some day when these instances are reliably available from @primitive@ we can
+-- likely replace this with 'Data.Primitive.ByteArray.ByteArray'.
+--
+module Codec.CBOR.ByteArray.Sliced
+  ( SlicedByteArray(..)
+  , fromByteString
+  , fromByteArray
+  , sizeofSlicedByteArray
+  , toByteString
+  , toBuilder
+  ) where
+
+import GHC.Exts
+import Data.Word
+import Foreign.Ptr
+import Control.Monad.ST
+import System.IO.Unsafe
+
+import qualified Data.Primitive.ByteArray as Prim
+import           Data.Primitive.Types (Addr(..))
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Unsafe as BS
+import qualified Data.ByteString.Short as BSS
+import qualified Data.ByteString.Short.Internal as BSS
+import qualified Data.ByteString.Builder as BSB
+import qualified Data.ByteString.Builder.Internal as BSB
+
+import Codec.CBOR.ByteArray.Internal
+
+data SlicedByteArray = SBA {unSBA :: !Prim.ByteArray, offset :: !Int, length :: !Int}
+
+fromByteString :: BS.ByteString -> SlicedByteArray
+fromByteString bs =
+    case BSS.toShort bs of
+      BSS.SBS ba -> fromByteArray (Prim.ByteArray ba)
+
+fromByteArray :: Prim.ByteArray -> SlicedByteArray
+fromByteArray ba = SBA ba 0 (Prim.sizeofByteArray ba)
+
+sizeofSlicedByteArray :: SlicedByteArray -> Int
+sizeofSlicedByteArray (SBA _ _ len) = len
+
+-- | Note that this may require a copy.
+toByteString :: SlicedByteArray -> BS.ByteString
+toByteString sba =
+    unsafePerformIO
+    $ BS.unsafePackCStringFinalizer ptr (sizeofSlicedByteArray sba) (touch pinned)
+  where
+    pinned = toPinned sba
+    !(Addr addr#) = Prim.byteArrayContents pinned
+    ptr = Ptr addr#
+
+toPinned :: SlicedByteArray -> Prim.ByteArray
+toPinned (SBA ba off len)
+  | isByteArrayPinned ba = ba
+  | otherwise = runST $ do
+        ba' <- Prim.newPinnedByteArray len
+        Prim.copyByteArray ba' 0 ba off len
+        Prim.unsafeFreezeByteArray ba'
+
+toBuilder :: SlicedByteArray -> BSB.Builder
+toBuilder = \(SBA ba off len) -> BSB.builder (go ba off len)
+  where
+    go ba !ip !ipe !k (BSB.BufferRange op ope)
+      | inpRemaining <= outRemaining = do
+          copyToAddr ba ip op inpRemaining
+          let !br' = BSB.BufferRange (op `plusPtr` inpRemaining) ope
+          k br'
+      | otherwise = do
+          copyToAddr ba ip op outRemaining
+          let !ip' = ip + outRemaining
+          return $ BSB.bufferFull 1 ope (go ba ip' ipe k)
+      where
+        outRemaining = ope `minusPtr` op
+        inpRemaining = ipe - ip
+
+instance IsList SlicedByteArray where
+  type Item SlicedByteArray = Word8
+  fromList xs = fromListN (Prelude.length xs) xs
+  fromListN n xs =
+      let arr = mkByteArray n xs
```

Comment:
What is the behaviour of `fromListN` supposed to be when `n` is different than the length of `xs`? Currently the result will be undetermined if `n` is less than `length xs` and writes out of bounds if it's the other way around. See `mkByteArray`.

Reply:
The documentation explicitly says that `fromListN` is not obligated to provide any particular behavior in this case,
> The hint can be used to  construct the structure @l@ more efficiently compared to 'fromList'. If the given hint does not equal to the input list's length the behaviour of 'fromListN' is not specified.

Reply:
I have reworked this to ensure a consistent result: we fill with zeros if `n > length xs` and truncate if `n < length xs`.